### DISCUSSION TITLE
cpptrace has a few rough edges with the build and deployment

### DISCUSSION THREAD

**Discussion by marlowa** [jeremy-rifkin/cpptrace#69] [General]
First of all, I have to say what an amazing bit of work cpptrace is! I just tried it out and it works a treat. Well done Jeremy for solving such a difficult problem and making it freely available. Thanks!
But I am trying to introduce its use into the corporation I work for and I note a few rough edges. First, the build tried to fetch things from git (dwarf-lite). That doesn't work in a corporate environment due to firewalls. So I had to build it at home. Second, I could not see a clean distribution directory coming out of the build process. The cpptrace.h header is in a different include dir to others that are also needed. I was hoping to see a distribution dir with include and lib underneath with all the stuff needed for full compiling and linking. All the stuff is present, it just needs marshalling. It's not a showstopper, I can do it by hand for the moment. But IMO it's something that needs a little attention. Finally, I note that on Red Hat the dwarf lib has just been deprecated. What implication, if any, does this have for cpptrace?

**Comment by jeremy-rifkin** [comment]
Thanks for your kind words, I'm happy to hear it worked well for you! I appreciate the feedback a lot; ease of use for build and deployment is a big priority.

Out of curiosity, would getting the package through conan or vcpkg be an option in your corporate environment? That might be the easiest way if fetchcontent isn't an option.

If building locally is the best way to go, would a normal `make install` be sufficient for a clean packaged build? It should put the .hpp files in the same include directory:
```
$ cmake .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=foo
...
$ make -j
...
$ make install
[ 76%] Built target dwarf-static
[100%] Built target cpptrace-lib
Install the project...
-- Install configuration: "Release"
-- Installing: C:/Users/rifkin/home/projects/cpptrace/build/foo/lib64/libdwarf.a
-- Installing: C:/Users/rifkin/home/projects/cpptrace/build/foo/lib/libdwarf.a
-- Installing: C:/Users/rifkin/home/projects/cpptrace/build/foo/lib/cmake/libdwarf/libdwarf-targets.cmake
-- Installing: C:/Users/rifkin/home/projects/cpptrace/build/foo/lib/cmake/libdwarf/libdwarf-targets-release.cmake
-- Installing: C:/Users/rifkin/home/projects/cpptrace/build/foo/lib/cmake/libdwarf/libdwarf-config.cmake
-- Installing: C:/Users/rifkin/home/projects/cpptrace/build/foo/include/libdwarf/libdwarf.h
-- Installing: C:/Users/rifkin/home/projects/cpptrace/build/foo/include/libdwarf/dwarf.h
-- Installing: C:/Users/rifkin/home/projects/cpptrace/build/foo/lib/pkgconfig/libdwarf.pc
-- Up-to-date: C:/Users/rifkin/home/projects/cpptrace/build/foo/include
-- Installing: C:/Users/rifkin/home/projects/cpptrace/build/foo/include/cpptrace
-- Installing: C:/Users/rifkin/home/projects/cpptrace/build/foo/include/cpptrace/cpptrace.hpp
-- Up-to-date: C:/Users/rifkin/home/projects/cpptrace/build/foo/include
-- Up-to-date: C:/Users/rifkin/home/projects/cpptrace/build/foo/include/cpptrace
-- Installing: C:/Users/rifkin/home/projects/cpptrace/build/foo/include/cpptrace/cpptrace_export.hpp
-- Installing: C:/Users/rifkin/home/projects/cpptrace/build/foo/lib/libcpptrace.a
-- Installing: C:/Users/rifkin/home/projects/cpptrace/build/foo/lib/cmake/cpptrace/cpptrace-config.cmake
-- Installing: C:/Users/rifkin/home/projects/cpptrace/build/foo/lib/cmake/cpptrace/cpptrace-config-version.cmake
-- Installing: C:/Users/rifkin/home/projects/cpptrace/build/foo/lib/cmake/cpptrace/cpptrace-targets.cmake
-- Installing: C:/Users/rifkin/home/projects/cpptrace/build/foo/lib/cmake/cpptrace/cpptrace-targets-release.cmake 
```
I am noticing that `lib/pkgconfig/libdwarf.pc` will have an absolute path baked into it but that can be fixed (or it might not be an issue depending on how you're using the library!).

Alternatively, libdwarf can be built from source and then `-DCPPTRACE_USE_EXTERNAL_LIBDWARF=On` can be used to have cpptrace use that rather than getting libdwarf through fetchcontent. I'll write out the steps below in case they are helpful (using `~/scratch/cpptrace-test` as my working path and installing to `~/scratch/cpptrace-test/resources`):

For building libdwarf from source:
```sh
cd ~/scratch/cpptrace-test
mkdir resources
git clone https://github.com/davea42/libdwarf-code.git
cd libdwarf-code
git checkout 6216e185863f41d6f19ab850caabfff7326020d7 # looks like some cmake stuff changed upstream, checking out the commit cpptrace currently uses
mkdir build
cd build
cmake .. -DPIC_ALWAYS=On -DBUILD_DWARFDUMP=Off -DCMAKE_INSTALL_PREFIX=~/scratch/cpptrace-test/resources
make -j
make install
# Consolidate compiler generated dependencies of target dwarf
# [100%] Built target dwarf
# Install the project...
# -- Install configuration: ""
# -- Installing: /homes/rifkin/scratch/cpptrace-test/resources/lib64/libdwarf.a
# -- Installing: /homes/rifkin/scratch/cpptrace-test/resources/lib/libdwarf.a
# -- Installing: /homes/rifkin/scratch/cpptrace-test/resources/lib/cmake/libdwarf/libdwarf-targets.cmake
# -- Installing: /homes/rifkin/scratch/cpptrace-test/resources/lib/cmake/libdwarf/libdwarf-targets-noconfig.cmake
# -- Installing: /homes/rifkin/scratch/cpptrace-test/resources/lib/cmake/libdwarf/libdwarf-config.cmake
# -- Installing: /homes/rifkin/scratch/cpptrace-test/resources/include/libdwarf/libdwarf.h
# -- Installing: /homes/rifkin/scratch/cpptrace-test/resources/include/libdwarf/dwarf.h
# -- Installing: /homes/rifkin/scratch/cpptrace-test/resources/lib/pkgconfig/libdwarf.pc
```

Then cpptrace:
```sh
cd ~/scratch/cpptrace-test
git clone https://github.com/jeremy-rifkin/cpptrace.git
cd cpptrace
git checkout v0.3.1
mkdir build
cd build
cmake .. -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=On -DCPPTRACE_USE_EXTERNAL_LIBDWARF=On -DCMAKE_PREFIX_PATH=~/scratch/cpptrace-test/resources -DCMAKE_INSTALL_PREFIX=~/scratch/cpptrace-test/resources
make -j
make install
# Consolidate compiler generated dependencies of target cpptrace-lib
# [100%] Built target cpptrace-lib
# Install the project...
# -- Install configuration: "Release"
# -- Up-to-date: /homes/rifkin/scratch/cpptrace-test/resources/include
# -- Installing: /homes/rifkin/scratch/cpptrace-test/resources/include/cpptrace
# -- Installing: /homes/rifkin/scratch/cpptrace-test/resources/include/cpptrace/cpptrace.hpp
# -- Up-to-date: /homes/rifkin/scratch/cpptrace-test/resources/include
# -- Up-to-date: /homes/rifkin/scratch/cpptrace-test/resources/include/cpptrace
# -- Installing: /homes/rifkin/scratch/cpptrace-test/resources/include/cpptrace/cpptrace_export.hpp
# -- Installing: /homes/rifkin/scratch/cpptrace-test/resources/lib/libcpptrace.so.0.3.1
# -- Installing: /homes/rifkin/scratch/cpptrace-test/resources/lib/libcpptrace.so.0
# -- Installing: /homes/rifkin/scratch/cpptrace-test/resources/lib/libcpptrace.so
# -- Installing: /homes/rifkin/scratch/cpptrace-test/resources/lib/cmake/cpptrace/cpptrace-config.cmake
# -- Installing: /homes/rifkin/scratch/cpptrace-test/resources/lib/cmake/cpptrace/cpptrace-config-version.cmake
# -- Installing: /homes/rifkin/scratch/cpptrace-test/resources/lib/cmake/cpptrace/cpptrace-targets.cmake
# -- Installing: /homes/rifkin/scratch/cpptrace-test/resources/lib/cmake/cpptrace/cpptrace-targets-release.c
```

Then everything should be good to go with the libraries located at `~/scratch/cpptrace-test/resources/[include, lib]`.

Regarding red hat deprecating libdwarf, I don't expect that to be an issue since cpptrace is getting the library through fetchcontent or a separate installation. I wouldn't expect libdwarf to just break on red hat, but maybe there's something red hat specific I don't know about.

**Reply by marlowa** [reply]
Thank you for those hints, I will do some more work.
In the meantime about your question: ""out of curiosity, would getting the package through conan or vcpkg be an option in your corporate environment? " - the answer is "no". I am working in a RHEL8 env where the only package management we have is yumming something in.  We can yum in libdwarf-devel  since that is in the RHEL repo, but it is not libdwarf-lite. And of course, cpptrace isn't there. I am also trying to get another github project into our system, fmt, and face similar struggles. It isn't in the RHEL repo either, but it does build nicely from source so I've made an arrangement with the CI/CD people for our builds to download a binary release from artifactory. I want to do something similar with cpptrace but it needs to build out the box easier first. A build that fetches from github is a no-go.

**Reply by marlowa** [reply]
I just tried version 0.5.4 and have found another rough edge. It now seems to depend on zstd but the cmake magic appears to try to download it if not found. The build error is in a corporate env where it fails trying to download zstd. Building on my own linux mint machine is fine.

**Reply by jeremy-rifkin** [reply]
Thanks, zstd is indeed needed. I'll try to update the instructions for a corporate environment.
